hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/subscription/SubscriptionOwnerCache.java

```diff
@@ -0,0 +1,82 @@
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.api.SubscriptionName;
+import pl.allegro.tech.hermes.api.TopicName;
+import pl.allegro.tech.hermes.domain.subscription.SubscriptionRepository;
+import pl.allegro.tech.hermes.management.domain.topic.TopicService;
+
+@Component
+public class SubscriptionOwnerCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(SubscriptionOwnerCache.class);
+
+    private final SubscriptionRepository subscriptionRepository;
+    private final TopicService topicService;
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    private Multimap<OwnerId, SubscriptionName> cache = Multimaps.synchronizedMultimap(ArrayListMultimap.create());
+
+    public SubscriptionOwnerCache(SubscriptionRepository subscriptionRepository, TopicService topicService,
+                                  @Value("${subscriptionOwnerCache.refreshRateInSeconds}") int refreshRateInSeconds) {
+        this.subscriptionRepository = subscriptionRepository;
+        this.topicService = topicService;
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("subscription-owner-cache-%d")
+                        .build());
+        scheduledExecutorService.scheduleAtFixedRate(this::refillCache, 0, refreshRateInSeconds, TimeUnit.SECONDS);
+    }
+
+    @PreDestroy
+    public void stop() {
+        scheduledExecutorService.shutdown();
+    }
+
+    Collection<SubscriptionName> get(OwnerId ownerId) {
+        return cache.get(ownerId);
+    }
+
+    void onRemovedSubscription(String subscriptionName, TopicName topicName) {
+        cache.entries().removeIf(entry -> entry.getValue().equals(new SubscriptionName(subscriptionName, topicName)));
+    }
+
+    void onCreatedSubscription(Subscription subscription) {
+        cache.put(subscription.getOwner(), subscription.getQualifiedName());
+    }
+
+    void onUpdatedSubscription(Subscription oldSubscription, Subscription newSubscription) {
+        cache.remove(oldSubscription.getOwner(), oldSubscription.getQualifiedName());
+        cache.put(newSubscription.getOwner(), newSubscription.getQualifiedName());
+    }
+
+    private void refillCache() {
+        try {
+            logger.info("Starting filling Owner Id to Subscription Name cache");
+            long start = System.currentTimeMillis();
+            Multimap<OwnerId, SubscriptionName> cache = ArrayListMultimap.create();
+            topicService.getAllTopics().stream()
+                    .flatMap(topic -> subscriptionRepository.listSubscriptions(topic.getName()).stream())
+                    .forEach(subscription -> cache.put(subscription.getOwner(), subscription.getQualifiedName()));
+            this.cache = Multimaps.synchronizedMultimap(cache);
+            long end = System.currentTimeMillis();
+            logger.info("Cache filled. Took {}ms", end - start);
```

Comment:
nitpicking: this log msg will be far far away from "Starting filling Owner Id....". So The question is "What cache?". I can see it only from logger name (class name), but it might be nice to have full statement here as well.

Reply:
done